### PR TITLE
Prompt documentation & fixes

### PR DESCRIPTION
**Pull request by Elv13**
I still have a bigger menubar refactor to add to the PR + fix luacheck and travis.

 * Add syntax highlight to the prompt
 * Add more timer documentation (unrelated, but whatever)
 * Add many prompt examples
 * Will eventually add a generic `awful.widget.common` based completion bar

### CODE REVIEW

**Comment by psychon** [comment]
> I still have a bigger menubar refactor to add to the PR + fix luacheck and travis.

You mean "I still have several other PRs to open so I do not create any more monster-PRs", right?

**Comment by Elv13** [comment]
> You mean "I still have several other PRs to open so I do not create any more monster-PRs", right?

As you wish. This PR doesn't have a lot of code, its all doc/tests and some minor things.

Here's the menubar core rewrite prototype https://gist.github.com/Elv13/bfe38dc8d8bc0a02a3ae273df9f65100 . It will need to be split in 2 so the awful.widget.common support is decoupled from the wibar code.

This PR is now ready

**Comment by Elv13** [comment]
@blueyed Can you review the prompt command hook example. It is inspired by a description of your system (I don't have the code, so I may be totally off). Also, there is a lot of content you have a lot of experience with, so I guess your well placed to review it.

About the decreased coverage, it should go away once the accuracy PR is merged. It is due to all the example tests callbacks not being called (there is no way to test them anyway, even if I improved the shims to support take inputs).

File: lib/awful/prompt.lua [hunk 1]
```diff
@@ -1,21 +1,52 @@
...
+--   <tr><td><kbd>CTRL+F</kbd></td><td></td></tr>
+--   <tr><td><kbd>CTRL+H</kbd></td><td>backward-delete-char</td></tr>
+--   <tr><td><kbd>CTRL+K</kbd></td><td>kill-line</td></tr>
+--   <tr><td><kbd>CTRL+U</kbd></td><td>unix-line-discard</td></tr>
+--   <tr><td><kbd>CTRL+W</kbd></td><td>unix-word-rubout</td></tr>
+--   <tr><td><kbd>CTRL+BACKSPACE</kbd></td><td>unix-word-rubout</td></tr>
+--   <tr><td><kbd>SHIFT+INSERT</kbd></td><td>paste</td></tr>
+--   <tr><td><kbd>HOME</kbd></td><td>beginning-of-line</td></tr>
+--   <tr><td><kbd>END</kbd></td><td>end-of-line</td></tr>
+-- </table>
```
Comment (blueyed) [inline lib/awful/prompt.lua]: Can this be a markdown table?

Comment (Elv13) [inline lib/awful/prompt.lua]: Discount supports table. Some time ago I tried to make ldoc display them properly, it didn't. This may be fixed (or not). Also, ldoc fallback to its own implementation if discount isn't installed. That one doesn't support table.

That being said, yes, I would like to use markdown there, but it has to work too...

File: tests/examples/wibox/awidget/prompt/highlight.lua [hunk 2]
```diff
@@ -0,0 +1,66 @@
+local parent    = ... --DOC_NO_USAGE --DOC_HIDE
+local wibox     = require( "wibox"     ) --DOC_HIDE
+local awful     = { prompt = require("awful.prompt") }--DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+
+    local amp = "&amp"..string.char(0x3B)
+    local quot = "&quot"..string.char(0x3B)
```
Comment (blueyed) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: Why is it indented?
Is it required for doc rendering?

Comment (Elv13) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: Yes, markdown requires 4 spaces for code. Some examples get away without it if they use `@usage`, but `@usage` is used differently for headers (as their documentation explains). This is why that example has a "DISABLE_USAGE" near the top. All lines with DOC_HIDE are hidden, so they don't need the indentation.

Also, before you ask, the ugly `string.char(0xFF)` is a bad hack around a bug. When I will rewrite this system to make @psychon happy, it wont use CMake, I can guarantee you that.

Comment (blueyed) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: Hasn't my tests-cleanup broke those then?!

What about having the examples in the source files, like Python doctests?
And they would be extracted and run from there?
I don't think it is bad to have the requires in there btw, since it makes it easier to adopt, doesn't it?

Comment (Elv13) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: At first I tried to have them in the doc, but ldoc isn't advanced enough to do that. Also, the little magic syntax to add the boilerplate is necessary, or it would require an external template for each test, making things worst instead of better.

I am not saying it is impossible to improve the current system. It is. What I am saying is that before saying "lets ditch it and start over", the replacement has to be *better* than the current solution.

> Hasn't my tests-cleanup broke those then?!

I hope not, some documentation pages are entirely written using it.

Once that "improve CI accuracy" PR is merged, I will open a new one to fix the other @psychon issue about the lines being off in users backtraces. This will also solve the speed issue as it wont run everytime cmake is called. The next steps would be to generate the test list using cmake, then have a single lua process generate all of them. I have a prototype of that, but it has its own share of issues.

As for having them "in" the lua file. It isn't impossible (and it is probably a good thing), but it will take a lot more work than you seem to think to get to the "status quo".

File: lib/gears/timer.lua [hunk 3]
```diff
@@ -105,6 +105,8 @@ local timer_instance_mt = {
 -- @tparam table args Arguments.
 -- @tparam number args.timeout Timeout in seconds (e.g. 1.5).
 -- @tparam[opt=false] boolean args.autostart Automatically start the timer.
+-- @tparam function args.callback Shortcut to avoid having to
+--  use `connect_signal`.
```
Comment (psychon) [inline lib/gears/timer.lua]: That is not really a good documentation. How about `Callback function to register. Equivalent to connecting to the "timeout" signal of the returned object.` or something like that.

File: lib/gears/timer.lua [hunk 4]
```diff
@@ -1,6 +1,47 @@
...
+--            widget       = wibox.widget.progressbar,
+--        },
+--        {
+--            id           = "mytb",
+--            text         = "100%",
+--            widget       = wibox.widget.textbox,
+--        },
+--        layout      = wibox.layout.stack,
+--        set_battery = function(self, val)
+--            self.mytb.text, self.mypb.value = tonumber(val).."%", tonumber(val)
```
Comment (psychon) [inline lib/gears/timer.lua]: I'd prefer this to be done in two lines instead of one. Easier to read.

File: lib/gears/timer.lua [hunk 5]
```diff
@@ -1,6 +1,47 @@
...
+--            -- instead of spawning a shell. This is only an example.
+--            awful.spawn.easy_async(
+--                {"sh", "-c", "acpi | sed -n 's/^.*, \([0-9]*\)%/\1/p'"},
+--                function(out)
+--                    mybatterybar.battery = out
+--                end
+--            )
+--        end
+--    }
+--
```
Comment (psychon) [inline lib/gears/timer.lua]: It's a nice example. :-)

For the commit message ("cannot be tested"):

I suggest adding an `:emit_signal("timeout")`. That way the widget is updated right at startup and half of this could be tested, I guess. Hm, perhaps not tested, but that `emit_signal` is still a good idea.

File: tests/examples/shims/_common_template.lua [hunk 6]
```diff
@@ -1,3 +1,4 @@
+-- luacheck: globals string
```
Comment (psychon) [inline tests/examples/shims/_common_template.lua]: Wrong fixup? This change should be in the previous commit.

File: lib/awful/prompt.lua [hunk 7]
```diff
@@ -485,16 +485,33 @@ function prompt.run(args, textbox, exe_callback, completion_callback,
...
                         local original_command = command
-                        if ret then
-                            command = ret
+
+                        -- Support both a "simple" and a "complex" way to
+                        -- control if the prompt should quit.
+                        quit = quit == nil and (ret ~= true) or (quit~=false)
+
+                        -- Allow the callback to change the command
+                        command = (ret ~= true) and ret or command
```
Comment (psychon) [inline lib/awful/prompt.lua]: Sigh, the joy of API compatibility. I don't really understand this and I'm lazy right now, so I'll let @blueyed decide. My only question is: API documentation?

File: lib/awful/prompt.lua [hunk 8]
```diff
@@ -485,16 +485,33 @@ function prompt.run(args, textbox, exe_callback, completion_callback,
...
+
+                        -- Quit by default, but allow it to be disabled
+                        if ret and type(ret) ~= "boolean" then
                             cb = exe_callback
-                        else
+                            if not quit then
+                                cur_pos = ret:wlen() + 1
```
Comment (psychon) [inline lib/awful/prompt.lua]: Why `wlen`? (Almost) nothing else here uses it. I guess you found some test case that requires this?

File: tests/examples/wibox/awidget/prompt/simple.lua [hunk 9]
```diff
@@ -2,6 +2,7 @@ local parent    = ... --DOC_NO_USAGE --DOC_HIDE
 local wibox     = require( "wibox"     ) --DOC_HIDE
 local awful     = { prompt = require("awful.prompt") }--DOC_HIDE
 local beautiful = require( "beautiful" ) --DOC_HIDE
+local naughty   = {} --DOC_HIDE
```
Comment (psychon) [inline tests/examples/wibox/awidget/prompt/simple.lua]: Unrelated change?

Comment (psychon) [inline tests/examples/wibox/awidget/prompt/simple.lua]: ...that should be in another commit

File: lib/awful/prompt.lua [hunk 10]
```diff
@@ -65,6 +65,13 @@
 --
 -- @DOC_wibox_awidget_prompt_hooks_EXAMPLE@
 --
+-- *[Example two] Modifiing the command (+ vi like input)*:
+--
+-- The hook system also allows to modify the command before interpreting it in
+-- the `exe_callback`.
+--
+-- @DOC_wibox_awidget_prompt_vilike_EXAMPLE@
+--
```
Comment (psychon) [inline lib/awful/prompt.lua]: Why "WIP" in the commit message?

File: lib/awful/prompt.lua [hunk 11]
```diff
@@ -65,6 +65,13 @@
 --
 -- @DOC_wibox_awidget_prompt_hooks_EXAMPLE@
 --
+-- *[Example two] Modifiing the command (+ vi like input)*:
```
Comment (psychon) [inline lib/awful/prompt.lua]: Less "i", please

File: tests/examples/wibox/awidget/prompt/vilike.lua [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local terminal  = "xterm"  --DOC_HIDE
+
+    local atextbox = wibox.widget.textbox()
+
+    -- Store a list of verbs characters in a hash
+    local verbs = {
+        -- Spawn in a terminal
+        t = function(adjs, count, cmd) return {terminal, "-e", cmd} end,  --luacheck: no unused args
+        -- Spawn width a shell
```
Comment (psychon) [inline tests/examples/wibox/awidget/prompt/vilike.lua]: width -> with

File: lib/awful/prompt.lua [hunk 13]
```diff
@@ -72,6 +72,16 @@
 --
 -- @DOC_wibox_awidget_prompt_vilike_EXAMPLE@
 --
+-- *[Example three] Key listener*:
+--
+-- The 2 previous examples were focused on changing the prompt behavior. This
+-- one explains or to "spy" on the prompt events. This can be used for
```
Comment (psychon) [inline lib/awful/prompt.lua]: Should the "or" be a "how"?

File: tests/examples/wibox/awidget/prompt/keypress.lua [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    awful.prompt.run {
+        prompt               = "<b>Run: </b>",
+        keypressed_callback  = function(mod, key, cmd) --luacheck: no unused args
+            if key == "Shift_L" then
+                notif = naughty.notify { text = "Shift pressed" }
+            end
+        end,
+        keyreleased_callback = function(mod, key, cmd) --luacheck: no unused args
+            if notif then
+                notif.destroy_cb(naughty.notificationClosedReason.undefined)
```
Comment (psychon) [inline tests/examples/wibox/awidget/prompt/keypress.lua]: Are you trying to destroy the notification? That's not how it works. This only called the `destroy = foo`-callback that you gave to `naughty.notify`, so should be `nil` here.

I think you want `naughty.destroy(notify)` (the reason argument can be nil).

File: lib/awful/prompt.lua (hunk 7 above)
Comment (Elv13) [inline lib/awful/prompt.lua]: > Sigh, the joy of API compatibility. 

Indeed, we entered the era of "don't break the API" again.

> My only question is: API documentation?

My nice example wasn't working ;) (the completion bar one posted in the gist). The first version of this PR had it (see the https://github.com/awesomeWM/awesome/pull/52 example), but it was already broken by the time the last revision was merged.

File: lib/awful/prompt.lua (hunk 8 above)
Comment (Elv13) [inline lib/awful/prompt.lua]: Right. To stay compliant with how the other code paths behave.

File: lib/gears/timer.lua (hunk 5 above)
Comment (Elv13) [inline lib/gears/timer.lua]: Ok, I will add more magic to be able to add shim commands to `$PATH`. That way we will have a stable output.

> It's a nice example. :-)

(as you know) coming from a genuine question. He was kind of right, there was bits of answer everywhere, but putting them together was non trivial.

The signal/connect part of the code is still undocumented (there is no `client:connect_signal` documentation because I have yet to create a shared version for CAPI, the widgets have one, but without an example)

**Comment by Elv13** [comment]
All done except the timer test. I propose to do that in another PR as it requires changes to the test framework[1] and you said this one is already big enough.

[1] To have stable `easy_async` output.

File: tests/examples/wibox/awidget/prompt/highlight.lua [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+        awful.prompt.run {
+            prompt       = "<b>Echo: </b>",
+            text         = 'a_very "nice" $SHELL && command', --DOC_HIDE
+            bg_cursor    = "#ff0000",
+            -- To use the default `rc.lua` prompt:
+            --textbox      = mouse.screen.mypromptbox.widget,
+            textbox      = atextbox,
+            highlighter  = function(b, a)
+                -- Add a random marker to delimitate the cursor
+                local cmd = b.."ZZZZZ"..a
```
Comment (psychon) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: By the way: Bad hack! (I just had to mention it, feel free to ignore this comment)

**Comment by psychon** [comment]
Pushed two fixup commits. Squash them and get @blueyed to say that https://github.com/awesomeWM/awesome/pull/1329/commits/e1a58dabff0d8c63d7a5cce03d37f18c95bb7fe2 is good and this is good to go in.

**Comment by blueyed** [comment]
e1a58da is gone, which one was it?

I will review it altogether, but not today anymore.

**Comment by psychon** [comment]
Heh. Your link works. Just in case, it points here: https://github.com/awesomeWM/awesome/commit/e1a58dabff0d8c63d7a5cce03d37f18c95bb7fe2

**Comment by Elv13** [comment]
@blueyed ping? (or just merge it)

The patch basically restore the ability not to exit. We discussed those use cases in Berlin last year.

**Comment by Elv13** [comment]
BUMP (again, sorry for being annoying)

File: lib/gears/timer.lua [hunk 16]
```diff
@@ -104,9 +146,14 @@ local timer_instance_mt = {
 --- Create a new timer object.
 -- @tparam table args Arguments.
 -- @tparam number args.timeout Timeout in seconds (e.g. 1.5).
+-- @tparam[opt=false] boolean args.autostart Automatically start the timer.
+-- @tparam[opt=nil] function args.callback Callback function to connect to the
+--  "timeout" signal.
+-- @tparam[opt=false] boolean args.single_shot Run only once then stop.
```
Comment (blueyed) [inline lib/gears/timer.lua]: Maybe `once` instead of `single_shot`?

File: tests/examples/shims/_common_template.lua (hunk 6 above)
Comment (blueyed) [inline tests/examples/shims/_common_template.lua]: @Elv13 
addressed?

File: tests/examples/wibox/awidget/prompt/highlight.lua (hunk 15 above)
Comment (blueyed) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: Yes, it should probably not be used in the docs like that.
Can we do better here?

File: tests/examples/wibox/awidget/prompt/vilike.lua [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+                    util   = require("awful.util")}--DOC_HIDE
+local beautiful = require( "beautiful" ) --DOC_HIDE
+local terminal  = "xterm"  --DOC_HIDE
+
+    local atextbox = wibox.widget.textbox()
+
+    -- Store a list of verbs characters in a hash
+    local verbs = {
+        -- Spawn in a terminal
+        t = function(adjs, count, cmd) return {terminal, "-e", cmd} end,  --luacheck: no unused args
```
Comment (blueyed) [inline tests/examples/wibox/awidget/prompt/vilike.lua]: It was mentioned somewhere else that the `-e` works differently on terminals.
Have we addressed this somehow already?

File: tests/examples/wibox/awidget/prompt/vilike.lua [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+        -- Spawn with a shell
+        s = function(adjs, count, cmd) return {awful.util.shell, '-c', cmd} end, --luacheck: no unused args
```
Comment (blueyed) [inline tests/examples/wibox/awidget/prompt/vilike.lua]: Might also behave differently wrt to quoting `cmd` or not.

File: tests/examples/wibox/awidget/prompt/highlight.lua (hunk 15 above)
Comment (Elv13) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: Any idea? Given the cursor is added to the text and the size of the content can be modified on both sides, there is not a million ways to "preserve" the cursor position and style. I agree it's ugly, but it's also simple and has a comment. Other solutions I can think of would either make the API worst or the example overcomplicated (but cleaner).

File: tests/examples/wibox/awidget/prompt/vilike.lua (hunk 17 above)
Comment (Elv13) [inline tests/examples/wibox/awidget/prompt/vilike.lua]: At one place, `terminal` was replaced with `urxvt`. This is not a solution, but could be done here too (or just ignore the problem and leave it as-is).

File: tests/examples/wibox/awidget/prompt/vilike.lua (hunk 18 above)
Comment (Elv13) [inline tests/examples/wibox/awidget/prompt/vilike.lua]: When? This is an argument array, not a parsed string. Both `bash` and `zsh` should handle this.

File: tests/examples/shims/_common_template.lua (hunk 6 above)
Comment (Elv13) [inline tests/examples/shims/_common_template.lua]: yes. GitHub doesn't check if the commit changed (as it would create too many problems). This was done not long after it was reported. This line is now part of `shims: Add the string.wlen function`

File: tests/examples/wibox/awidget/prompt/highlight.lua (hunk 15 above)
Comment (psychon) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: Only suggestion that I have to offer is to treat both parts separately. That means less highlight in some cases, but also less hacks...?

Comment (Elv13) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: Trading a mild hack for an obvious bug doesn't seems like such a good idea to me. Maybe I should simply replace `ZZZZZ` by a better name (like `ZZZCURSORZZZ`[1])?

[1] I prefer to limit it to a single char category (like upper caps) to avoid too many of the parser expressions breakage. `ZZZ` because it has the lowest risk of collision (beside a PIZZZA typo).

Comment (blueyed) [inline tests/examples/wibox/awidget/prompt/highlight.lua]: `ZZZCURSORZZZ` seems good to me then.

File: lib/gears/timer.lua (hunk 16 above)
Comment (Elv13) [inline lib/gears/timer.lua]: Qt call it `singleShot` and glib doesn't have it. My naming process is usually glib->qt->random. `once` isn't a bad name, but so is `single_shot`. If you *really* want me to change it, I will, but right now I left it as it is.

**Comment by Elv13** [comment]
Ok, rebased again with the ZZZZZ change. I think I addressed most issues. The last one is the terminal `-e`. There is much to do for this one. It *is* a problem, but there is no obvious solution beside extending `awful.spawn` to support it. This is out-of-scope for this PR. I could add a comment warning users about the issue, or just ignore it.

**Comment by blueyed** [comment]
@Elv13 
👍
